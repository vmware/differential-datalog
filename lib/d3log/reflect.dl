/* Reflect the contents of all distributed relations and streams
 * back to DDlog, so that the D3log runtime can manipulate it
 * (buffer, route, forward, etc).
 *
 * Relations declared here are populated by compiler-generated rules
 * when the compiler is run with `--d3log-dev` flag. */

import ddlog_std
import internment

// Rows from distributed relations.
//
// `relname`     - relation name
// `fact`        - row in the relation
// `destination` - destination specified via @-annotation, if any
relation DistributedRelFacts(
    relname: istring,
    fact: Any,
    destination: Option<D3logLocationId>
)

// Rows from distributed streams.
//
// `relname`     - stream name
// `fact`        - record in the stream
// `destination` - destination specified via @-annotation, if any
stream DistributedStreamFacts(
    relname: istring,
    fact: Any,
    destination: Option<D3logLocationId>
)
