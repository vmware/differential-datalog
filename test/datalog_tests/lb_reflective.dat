start;

insert lb::Data(100),
insert lb::Data(200),
insert lb::Data(300),
insert lb::Workers(1),

commit dump_changes;

start;

insert lb::Workers(2),

commit dump_changes;
