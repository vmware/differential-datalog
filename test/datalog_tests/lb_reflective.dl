import lb
import d3log::reflect

output relation OutputRelFacts(
    relname: istring,
    fact: Any,
    destination: Option<D3logLocationId>)

output stream OutputStreamFacts(
    relname: istring,
    fact: Any,
    destination: Option<D3logLocationId>)

OutputRelFacts(relname, fact, destination) :- DistributedRelFacts(relname, fact, destination).
OutputStreamFacts(relname, fact, destination) :- DistributedStreamFacts(relname, fact, destination).
