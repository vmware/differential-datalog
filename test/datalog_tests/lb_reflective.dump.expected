OutputRelFacts:
OutputRelFacts{.relname = "lb::BalancedData", .fact = lb::BalancedData{.dat = 100}, .destination = ddlog_std::Some{.x = 1}}: +1
OutputRelFacts{.relname = "lb::BalancedData", .fact = lb::BalancedData{.dat = 200}, .destination = ddlog_std::Some{.x = 1}}: +1
OutputRelFacts{.relname = "lb::BalancedData", .fact = lb::BalancedData{.dat = 300}, .destination = ddlog_std::Some{.x = 1}}: +1
OutputRelFacts:
OutputRelFacts{.relname = "lb::BalancedData", .fact = lb::BalancedData{.dat = 100}, .destination = ddlog_std::Some{.x = 1}}: -1
OutputRelFacts{.relname = "lb::BalancedData", .fact = lb::BalancedData{.dat = 100}, .destination = ddlog_std::Some{.x = 2}}: +1
OutputRelFacts{.relname = "lb::BalancedData", .fact = lb::BalancedData{.dat = 200}, .destination = ddlog_std::Some{.x = 1}}: -1
OutputRelFacts{.relname = "lb::BalancedData", .fact = lb::BalancedData{.dat = 200}, .destination = ddlog_std::Some{.x = 2}}: +1
